text/0000-layout-packed-aligned.md

```diff
@@ -0,0 +1,176 @@
+}
+```
+
+This is commonly needed when Rust is being used to interop with existing C and C++ code bases, which may contain
+unaligned types. For example in `clang` it is possible to create the following type definition, and there is
+currently no easy way to create a matching Rust type:
+
+```cpp
+struct  __attribute__((packed, aligned(4))) MyStruct {
+    uint8_t a;
+    uint32_t b;
+};
+```
+
+Currently `#[repr(packed(_))]` structs cannot transitively contain #[repr(align(_))] structs due to differing behavior between msvc and gcc/clang.
+However, in most cases, the user would expect `#[repr(C)]` to produce a struct layout matching the same type as defined by the current target. 
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+## `#[repr(C)]`
+When `align` and `packed` attributes exist on the same type, or when `packed` structs transitively contains `align` types,
+the resulting layout matches the target toolchain ABI.
+
+For example, given:
+```c
+#[repr(C, align(4))]
+struct Foo(u8);
+#[repr(C, packed(1))]
+struct Bar(Foo);
+```
+`align_of::<Bar>()` would be 4 for `*-pc-windows-msvc` and 1 for everything else.
+
+
+## `#[repr(system)]`
+When `align` and `packed` attributes exist on the same type, or when `packed` structs transitively contains `align` types,
+the resulting layout matches the target OS ABI.
+
+For example, given:
+```c
+#[repr(system, align(4))]
+struct Foo(u8);
+#[repr(system, packed(1))]
+struct Bar(Foo);
+```
+`align_of::<Bar>()` would be 4 for `*-pc-windows-msvc` and `*-pc-windows-gnu`. It would be 1 for everything else.
+
+
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+In the following paragraphs, "Decreasing M to N" means:
+```
+if M > N {
+    M = n
+}
+```
+
+"Increasing M to N" means:
+```
+if M < N {
+    M = N
+}
+```
+
+
+`#[repr(align(N))]` increases the base alignment of a type to be N.
+
+`#[repr(packed(M))]` decreases the alignment of the struct fields to be M. Because the base alignment of the type
+is defined as the maximum of the alignment for any fields, this also has the indirect result of decreasing the base
+alignment of the type to be M.
+
+When the align and packed modifiers are applied on the same type as `#[repr(align(N), packed(M))]`,
+the alignment of the struct fields are decreased to be M. Then, the base alignment of the type is
+increased to be N.
+
+When a `#[repr(packed(M))]` struct transitively contains a field with `#[repr(align(N))]` type,
+- The field is first `pad_to_align`. Then, the field is added to the struct with alignment decreased to M. The packing requirement overrides the alignment requirement. (GCC, `#[repr(Rust)]`, `#[repr(C)]` on gnu targets, `#[repr(system)]` on non-windows targets)
+- The field is added to the struct with alignment increased to N. The alignment requirement overrides the packing requirement. (MSVC, `#[repr(C)]` on msvc targets, `#[repr(system)]` on windows targets)
```

Comment:
We're not introducing a new type of alignment. Conceptually:

- `packed(N)` changes the alignment of the inner fields to be `N`.
- `aligned(M)` changes the alignment of the entire structure to be `M`.

When both rules apply to the same type, (that's when a packed type transitively contains an aligned field), we have ambiguity, and we need some rules to determine who wins.


So really we're just changing the value returned by `offset_of`. `align_of` will continue to return the base alignment of a type like before.

Reply:
But it is a different kind of alignment. `i64` has an alignment of 8, which can be lowered by packed. `#[repr(align(8))] i64` has an alignment of 8, which cannot be lowered by packed. `#[repr(align(4))] i64` has an alignment of 8, which can be lowered to 4 by packed. What is this if not a new kind of alignment? Attributes apply to the type they decorate, not to all types which transitively contain the decorated type.

Reply:
To reiterate this point: you end up with "primitive alignment" which can be suppressed by `packed`, originating from primitives; as well as "user alignment" which cannot be overridden by `packed`, originating from source attributes.

Types are not constructed as a list of primitive fields flattened from the relevant types, even as much as the MSVC layout algorithm sometimes pretends that this is the case. Types are instead defined compositionally, and a field of a struct or enum type is no different from that of some primitive type.

Reply:
And here "user alignment" would be what is ultimately returned by `align_of`, i.e. the alignment required for references to this type?

This reminds me a lot of the discussions around `u64` on 32bit MSVC, which requires separating "alignment of references to this type" and "alignment of struct fields to this type"... is the new kind of alignment proposed in this RFC the same as as that, or does MSVC actually have 3 kinds of alignment?

Reply:
Well I think the "alignment lowerbility" is only a concept relevant to the specific case of nested aligned types in #[repr(C)] structs; introducing a brand new user-queryable `std::mem::align_attribute` seems a bit overkill. The only thing that the user might be interested in would be the offset of the corresponding field within the packed structure. Whether the alignment can be overriden by `packed` is an implementation detail of `packed` that the user should not have to care about.

Reply:
Whether that new kind of alignment can be queried by the user is a secondary question. But if this concept exists it needs to be named and discussed in the RFC and the Rust Reference, and will become part of the models of Rust that formally describe its behavior.

This overlaps with my [comment over here](https://github.com/rust-lang/rfcs/pull/3718#discussion_r1824936376) saying that the algorithm as you currently describe it is quite hard to understand.

Also `align_attribute` seems like an odd name, it's not the attribute that you need here. It's the fact that types now have two kinds of alignment, that need to have distinct names and that are used in distinct situations. I don't fully understand it yet, that's why I'd like you to explain it in the RFC. :D

Reply:
I think any new kind of alignment should be exposed to user code specifically because user code may want to have its own layout algorithm (e.g. for JIT). Maybe add it as a new field to `Layout` and add methods for msvc struct layout?

Reply:
Adding fields to `core::alloc::Layout` will be a difficult sell, because `Layout` is part of the global dynamic allocation ABI, and adding more fields to allocation queries which are not needed for the actual allocation wouldn't be very zero cost abstraction of us.

Reply:
actually, now that I'm thinking about it, you'd probably want `repr(MS)` to be available on non-windows too since you'd need it in writing software like [Wine](https://www.winehq.org/) where you have to interface with both the host system and with Windows programs running inside your Win32 API implementation